contracts/stake-cw20-external-rewards/src/contract.rs

```diff
@@ -0,0 +1,1324 @@
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cosmwasm_std::{
+    from_binary, to_binary, Addr, BankMsg, Binary, Coin, CosmosMsg, Deps, DepsMut, Env,
+    MessageInfo, QuerierWrapper, Response, StdResult, Timestamp, Uint128, WasmMsg,
+};
+use cw20::{Cw20ExecuteMsg, Cw20ReceiveMsg, Denom};
+use std::cmp::min;
+
+use crate::error::ContractError;
+use crate::error::ContractError::RewardsNotStarted;
+use crate::msg::{
+    ClaimableRewardsResponse, ExecuteMsg, InfoResponse, InstantiateMsg, QueryMsg, ReceiveMsg,
+};
+use crate::state::{Config, LastClaim, CONFIG, LAST_CLAIMED};
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    validate_instantiate_msg(&env, &msg)?;
+    let config = Config {
+        start_block: msg.start_block,
+        end_block: msg.end_block,
+        payment_per_block: msg.payment_per_block,
+        total_amount: msg.total_amount,
+        denom: msg.denom.clone(),
+        staking_contract: deps.api.addr_validate(&*msg.distribution_token)?,
+        funded: match msg.denom {
+            Denom::Native(_) => true,
+            Denom::Cw20(_) => false,
+        },
+        payment_block_delta: msg.payment_block_delta,
+    };
+    CONFIG.save(deps.storage, &config)?;
+    Ok(Response::default())
+}
+
+pub fn validate_instantiate_msg(env: &Env, msg: &InstantiateMsg) -> Result<(), ContractError> {
+    if env.block.height > msg.start_block {
+        return Err(ContractError::StartBlockAlreadyOccurred {});
+    }
+    if msg.start_block > msg.end_block {
+        return Err(ContractError::StartBlockAfterEndBlock {});
+    }
+    if msg.start_block % msg.payment_block_delta != 0 {
+        return Err(ContractError::StartBlockNotDivisibleByPaymentDelta {});
+    }
+    if msg.end_block % msg.payment_block_delta != 0 {
+        return Err(ContractError::EndBlockNotDivisibleByPaymentDelta {});
+    }
+    let duration = Uint128::from(((msg.end_block - msg.start_block) / msg.payment_block_delta) + 1);
+    let calculated_total = duration
+        .checked_mul(msg.payment_per_block)
+        .map_err(cosmwasm_std::StdError::overflow)?
+        .checked_mul(Uint128::from(msg.payment_block_delta))
+        .map_err(cosmwasm_std::StdError::overflow)?;
```

Comment:
Should we really multiply by `payment_block_delta` here? Seems like `calculated_total = payment_per_block * duration` is what we want.

Reply:
ah, I see. I was kinda a dum dum here. I need to not divide by payment_block_delta when caculating duration, then I do not need to multiply by it later.